Parallel 2.0 dev

lunarblock opened the pull request:
### Description

add a description of your changes here...

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

yutianwu commented on core/state/journal.go:
```diff
@@ -166,7 +166,12 @@ func (ch resetObjectChange) dirtied() *common.Address {
 func (ch suicideChange) revert(s *StateDB) {
-	obj := s.getStateObject(*ch.account)
+	var obj *StateObject
+	if s.parallel.isSlotDB {
```
we should abstract this into one function

brilliant-lx replied:
Yes, I will try to remove this code.

setunapo replied:
Yes, I will try to remove this code.

yutianwu commented on core/state/state_object.go:
```diff
@@ -284,7 +327,8 @@ func (s *StateObject) GetCommittedState(db Database, key common.Hash) common.Has
 			meter = &s.db.StorageReads
 		}
-		if enc, err = s.getTrie(db).TryGet(key.Bytes()); err != nil {
+		if enc, err = s.getTrie(db).TryGet(key.Bytes()); err != nil { // fixme: handle trie concurrent safe
```
that‘s fine , we will disable parallel mode if snap is nil

yutianwu commented on core/state/state_object.go:
```diff
@@ -214,7 +252,7 @@ func (s *StateObject) getOriginStorage(key common.Hash) (common.Hash, bool) {
 			return common.Hash{}, false
 		}
-		s.originStorage[key] = val.(common.Hash)
+		// s.originStorage[key] = val.(common.Hash)
```
why do we comment this

setunapo replied:
there will be a concurrent issue.

yutianwu commented on core/state/state_object.go:
```diff
@@ -214,7 +252,7 @@ func (s *StateObject) getOriginStorage(key common.Hash) (common.Hash, bool) {
 			return common.Hash{}, false
 		}
-		s.originStorage[key] = val.(common.Hash)
+		// s.originStorage[key] = val.(common.Hash)
```
why delete this

yutianwu commented on core/state/state_object.go:
```diff
@@ -198,10 +202,44 @@ func (s *StateObject) GetState(db Database, key common.Hash) common.Hash {
+		return value
+	}
+	// on merge
+	value, dirty = s.pendingStorage[key]
```
In which case we will read states in `pendingStorage`

yutianwu commented on core/state/statedb.go:
```diff
@@ -711,54 +985,127 @@ func (s *StateDB) SystemAddressRedo() bool {
+				"addr", addr)
+			// 1.Try to get from dirty
+			if val, ok := obj.dirtyStorage[hash]; ok {
+				log.Info("GetState key in dirty", "SlotIndex", s.parallel.SlotIndex, "txIndex", s.txIndex,
```
don't think directly read `dirtyStorage` and `pendingStorage` from StateObject is a good idea in StateDB

yutianwu commented on core/state/statedb.go:
```diff
@@ -711,54 +985,127 @@ func (s *StateDB) SystemAddressRedo() bool {
+				return val
+			}
+			// 2.Try to get from pending, it is for merge to get KV from finalized StateObject.
+			if val, ok := obj.pendingStorage[hash]; ok {
```
in which case we will read `pendingStorage`

yutianwu commented on core/state/statedb.go:
```diff
@@ -834,25 +1193,53 @@ func (s *StateDB) StorageTrie(addr common.Address) Trie {
+func (s *StateDB) unconfirmedLightCopy(obj *StateObject) *StateObject {
+	newObj := obj.lightCopy(s)
+	// do balance fixup
+	if balance := s.getBalanceFromUnconfirmedDB(obj.address); balance != nil {
```
```
		newStateObject := s.unconfirmedLightCopy(stateObject)
		newStateObject.AddBalance(amount)
		s.parallel.dirtiedStateObjectsInSlot[addr] = newStateObject
```

since we will set `dirtiedStateObjectsInSlot` after `unconfirmedLightCopy`, why `getBalanceFromUnconfirmedDB` since it will always return nil.

`if obj, exist := db.parallel.dirtiedStateObjectsInSlot[addr]; exist {`

yutianwu commented on core/state/statedb.go:
```diff
@@ -635,39 +672,269 @@ func (s *StateDB) SubRefund(gas uint64) {
+			if obj, ok := db.parallel.dirtiedStateObjectsInSlot[addr]; ok { // if deleted on merge, can get from main StateDB, ok but fixme: concurrent safe
+				if _, ok := db.parallel.stateChangesInSlot[addr]; ok {
+					// fixme: dirtyStorage will be cleared on StateObject.finalize, to keep the KV in stateChangesInSlot?
+					if val, exist := obj.dirtyStorage[key]; exist {
```
we also should avoid directly using the `dirtyStorage` and `pendingStorage`. we can abstract these to a function like `obj.GetDirtiedState()`

yutianwu commented on core/state/statedb.go:
```diff
@@ -114,27 +115,39 @@ type SlotChangeList struct {
+	BalanceChangeReferred map[common.Address]*big.Int
+	CodeChangeReferred    map[common.Address]struct{} // fixme: codehash to compare? it may not necessary
+	// NonceChangeSet      map[common.Address]struct{}  // nonce should be reliable, it is not necessary
+	AddrStateReferred map[common.Address]struct{} // the address of StateObject we refer, or the state change?
```
is there any possibility that we add an address to  `AddrStateReferred` and then we get the changed code but not add it to `CodeChangeReferred`.  or do we check it in the conflict check.

yutianwu commented on core/state/statedb.go:
```diff
@@ -635,39 +672,269 @@ func (s *StateDB) SubRefund(gas uint64) {
+}
+
+// for addr state check of: Exist(), Empty() and HasSuicided()
+func (s *StateDB) getAddrStateFromUnconfirmedDB(addr common.Address) *StateObject {
```
is there any difference between `getStateObjectFromUnconfirmedDB` and `getAddrStateFromUnconfirmedDB`

yutianwu commented on core/state/statedb.go:
```diff
@@ -635,39 +672,269 @@ func (s *StateDB) SubRefund(gas uint64) {
+					KVChangeReferred:      make(map[common.Address]Storage),
+				}
+			}
+			if obj, ok := db.parallel.dirtiedStateObjectsInSlot[addr]; ok {
```
why check `addrStateChangesInSlot` here

yutianwu commented on core/state/statedb.go:
```diff
@@ -635,39 +672,269 @@ func (s *StateDB) SubRefund(gas uint64) {
+					KVChangeReferred:      make(map[common.Address]Storage),  // the KV pair should be exist
+				}
+			}
+			if obj, exist := db.parallel.dirtiedStateObjectsInSlot[addr]; exist {
```
why check `addrStateChangesInSlot` here

yutianwu commented on core/state/state_object.go:
```diff
@@ -268,9 +308,12 @@ func (s *StateObject) GetCommittedState(db Database, key common.Hash) common.Has
 		//      have been handles via pendingStorage above.
 		//   2) we don't have new values, and can deliver empty response back
-		if _, destructed := s.db.snapDestructs[s.address]; destructed {
+		s.db.snapParallelLock.RLock()
```
maybe we need to check if we are in a slotdb before adding a lock

yutianwu commented on core/state/state_object.go:
```diff
@@ -198,10 +202,44 @@ func (s *StateObject) GetState(db Database, key common.Hash) common.Hash {
+	// fixme: if s is light copied, access mainStateDB's stateobject's dirty, if lightCopy's dirty missed
+	val := s.GetCommittedState(db, key)
+	var mainVal common.Hash
+	if !s.inMainDB {
```
as discussed, we can come up with a better way to replace `inMainDB`

yutianwu commented on core/state/state_object.go:
```diff
@@ -198,10 +202,44 @@ func (s *StateObject) GetState(db Database, key common.Hash) common.Hash {
+			log.Info("StateObject::GetState loadStateObj")
+			// obj = mainObj
+			mainVal = mainObj.GetCommittedState(db, key)
+			if mainVal != val {
```
a bit confusing here why we need to check the committed state here

lunarblock committed:
[WIP] the Implementaion of Parallel EVM 2.0

1.features of 2.0:
  ** Streaming Pipeline
  ** Implement universal unconfirmed state db reference, try best to get account object state.
  ** New conflict detect, check based on what it has read.
  ** Do parallel KV conflict check for large KV read
  ** new Interface StateDBer and ParallelStateDB
  ** shared memory pool for parallel objects
  ** use map in sequential mode and sync.map in parallel mode for concurrent StateObject access
  ** replace DeepCopy by LightCopy to avoid redundant memory copy of StateObject
  ** do trie prefetch in advance
  ** dispatcher 2.0
     Static Dispatch & Dynamic Dispatch
     Stolen mode for TxReq when a slot finished its static dispatched tasks
     RealTime result confirm in Stage2, when most if the tx have been executed at least once
     Make it configurable

2.Handle of corner case:
 ** don't panic if there is anything wrong reading state
 ** handle system address, skip its balance check
 ** handle WBNB contract to reduce conflict rate by balance make up
    WBNB balance makeup by GetBalanceOpCode & depth
    add a lock to fix WBNB make up concurrent crash
    add a new interface GetBalanceOpCode

forcodedancing committed:
improve snap mutex
